src/http-strategies/Koa.js

```diff
@@ -0,0 +1,220 @@
+import Q from "q";
+import vary from "vary";
+import contentType from "content-type";
+import getRawBody from "raw-body";
+import API from "../controllers/API";
+import APIError from "../types/APIError";
+import Request from "../types/HTTP/Request";
+
+/**
+ * This controller receives requests directly from Koa and sends responses
+ * direclty through it, but it converts incoming requests to, and generates
+ * responses, from Request and Response objects that are defined by this
+ * framework in a way that's not particular to Koa. This controller thereby
+ * acts as a translation-layer between Koa and the rest of this json-api
+ * framework.
+ *
+ * @param {Object} options A set of configuration options.
+ *
+ * @param {boolean} options.tunnel Whether to turn on PATCH tunneling. See:
+ *    http://jsonapi.org/recommendations/#patchless-clients
+ *
+ * @param {boolean} options.handleContentNegotiation If the JSON API library
+ *    can't produce a representation for the response that the client can
+ *    `Accept`, should it return 406 or should it hand the request back to
+ *    Koa (i.e. yield next) so that subsequent handlers can attempt to
+ *    find an alternate representation? By default, it does the former. But you
+ *    can set this option to false to have this code just pass on to express.
+ */
+export default class KoaStrategy {
+  constructor(apiController, docsController, options) {
+    const defaultOptions = {
+      tunnel: false,
+      handleContentNegotiation: true
+    };
+
+    this.api = apiController;
+    this.docs = docsController;
+    this.config = Object.assign({}, defaultOptions, options); // apply options
+  }
+
+  // For requests like GET /:type, GET /:type/:id/:relationship,
+  // POST /:type PATCH /:type/:id, PATCH /:type, DELETE /:type/:idOrLabel,
+  // DELETE /:type, GET /:type/:id/links/:relationship,
+  // PATCH /:type/:id/links/:relationship, POST /:type/:id/links/:relationship,
+  // and DELETE /:type/:id/links/:relationship.
+  apiRequest() {
+    const strategy = this;
+    return function *(next){
+      try {
+        const reqObj = yield buildRequestObject(this, strategy.config.tunnel);
+        const resObj = yield strategy.api.handle(reqObj, this);
+        const skip = strategy.sendResources(resObj, this);
+        if(skip){
+          yield next;
+        }
+      }
+      catch (err) {
+        strategy.sendError(err, this);
+      }
+    };
+  }
+
+  // For requests for the documentation.
+  docsRequest() {
+    const strategy = this;
+    return function *(next){
+      try {
+        const reqObj = yield buildRequestObject(this, strategy.config.tunnel);
+        const resObj = yield strategy.docs.handle(reqObj, this);
+        const skip = strategy.sendResources(resObj, this);
```

Comment:
Can we name this variable something a bit more descriptive than `skip`? Maybe `delegate406Handling` or similar?